File: crates/raven-adapters/src/in_memory/workloads.rs
use std::{collections::VecDeque, fmt};

use rand::{RngExt, SeedableRng};
use raven_core::types::{FloatScalar, Strict, StrictCarrierOps};
use rustc_hash::{FxBuildHasher, FxHashMap};

use super::{InMemoryGraphError, InMemoryUndirectedGraph};

#[derive(Debug, Clone, PartialEq)]
pub struct GeneratedSbmCommands {
    pub nodes: Vec<usize>,
    pub operations: Vec<SbmInstruction>,
    pub expected_edges: usize,
    pub cluster_labels: Vec<usize>,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum SbmInstruction {
    Insert {
        time: i64,
        u: usize,
        v: usize,
        weight_delta: f64,
    },
    Delete {
        time: i64,
        u: usize,
        v: usize,
    },
    SetWeight {
        time: i64,
        u: usize,
        v: usize,
        weight: f64,
    },
}

impl SbmInstruction {
    pub fn time(self) -> i64 {
        match self {
            Self::Insert { time, .. }
            | Self::Delete { time, .. }
            | Self::SetWeight { time, .. } => time,
        }
    }
}

#[derive(Debug, Clone)]
pub struct SbmDiffWorkload<T> {
    pub nodes: Vec<usize>,
    pub cluster_labels: Vec<usize>,
    pub expected_edges: usize,
    pub snapshot_step: i64,
    pub batches: Vec<SbmUpdateBatch<T>>,
}

#[derive(Debug, Clone)]
pub struct SbmUpdateBatch<T> {
    pub time: i64,
    pub edge_ops: Vec<SbmEdgeOp<T>>,
    pub node_ops: Vec<(usize, Option<Strict<T>>)>,
}

impl<T> SbmUpdateBatch<T>
where
    T: FloatScalar,
    Strict<T>: StrictCarrierOps<Scalar = T> + Copy,
{
    pub fn apply_to_graph(
        &self,
        graph: &mut InMemoryUndirectedGraph<usize, T>,
    ) -> Result<(), InMemoryGraphError> {
        for op in &self.edge_ops {
            op.apply_to_graph(graph)?;
        }
        Ok(())
    }

    pub fn apply_to_graph_and_flush_node_ops(
        &self,
        graph: &mut InMemoryUndirectedGraph<usize, T>,
    ) -> Result<Vec<(usize, Option<Strict<T>>)>, InMemoryGraphError> {
        self.apply_to_graph(graph)?;
        Ok(graph.flush_node_ops())
    }
}

#[derive(Debug, Clone, Copy)]
pub enum SbmEdgeOp<T> {
    Set {
        u: usize,
        v: usize,
        weight: Strict<T>,
    },
    Delete {
        u: usize,
        v: usize,
    },
}

impl<T> SbmEdgeOp<T>
where
    T: FloatScalar,
    Strict<T>: StrictCarrierOps<Scalar = T> + Copy,
{
    pub fn apply_to_graph(
        self,
        graph: &mut InMemoryUndirectedGraph<usize, T>,
    ) -> Result<(), InMemoryGraphError> {
        match self {
            Self::Set { u, v, weight } => graph.update_edge(u, v, Some(weight)),
            Self::Delete { u, v } => match graph.update_edge(u, v, None) {
                Ok(()) | Err(InMemoryGraphError::MissingEdge) => Ok(()),
                Err(err) => Err(err),
            },
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SbmWorkloadError {
    InvalidParameter(&'static str),
    InvalidWeight,
    Graph(InMemoryGraphError),
}

impl fmt::Display for SbmWorkloadError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::InvalidParameter(param) => write!(f, "invalid SBM workload parameter: {param}"),
            Self::InvalidWeight => write!(f, "generated edge weight was not positive and finite"),
            Self::Graph(err) => write!(f, "failed to replay generated graph operation: {err}"),
        }
    }
}

impl std::error::Error for SbmWorkloadError {}

impl From<InMemoryGraphError> for SbmWorkloadError {
    fn from(value: InMemoryGraphError) -> Self {
        Self::Graph(value)
    }
}

/// Generate an SBM-style stream of edge updates with finite lifetimes.
///
/// This is a direct Raven-friendly port of the generator used by dyn-cc's
/// `prepare_diff_workload_sbm`: it samples internal vs external endpoints from
/// the expected SBM edge mass, increments edge weights by one, and expires each
/// increment after a fixed lifetime.
pub fn generate_sbm_commands(
    seed: u64,
    n_per_cluster: usize,
    k_clusters: usize,
    p_internal: f64,
    q_external: f64,
    n_multiplier: usize,
    lifetime_multiplier: f64,
) -> Result<GeneratedSbmCommands, SbmWorkloadError> {
    validate_sbm_params(
        n_per_cluster,
        k_clusters,
        p_internal,
        q_external,
        n_multiplier,
        lifetime_multiplier,
        1.0,
    )?;

    let mut rng = rand::rngs::StdRng::seed_from_u64(seed);

    let total_nodes = n_per_cluster * k_clusters;
    let nodes = (0..total_nodes).collect::<Vec<_>>();

    let cluster_nodes = (0..k_clusters)
        .map(|c| {
            let start = c * n_per_cluster;
            (start..start + n_per_cluster).collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();

    let expected_internal = (k_clusters as f64)
        * (n_per_cluster as f64)
        * ((n_per_cluster - 1) as f64)
        * 0.5
        * p_internal;
    let expected_external = (k_clusters * (k_clusters - 1)) as f64
        * 0.5
        * (n_per_cluster * n_per_cluster) as f64
        * q_external;
    let expected_edges = expected_internal + expected_external;
    if !(expected_edges.is_finite() && expected_edges > 0.0) {
        return Err(SbmWorkloadError::InvalidParameter(
            "expected edge count must be positive",
        ));
    }

    let num_updates = ((n_multiplier as f64) * expected_edges).ceil() as usize;
    let lifetime_steps = ((lifetime_multiplier * expected_edges).ceil() as i64).max(1);
    let internal_prob = expected_internal / expected_edges;

    let mut edge_weights = FxHashMap::<(usize, usize), f64>::with_capacity_and_hasher(
        expected_edges.ceil() as usize,
        FxBuildHasher,
    );
    let mut expirations = VecDeque::<(i64, (usize, usize))>::new();
    let mut operations = Vec::with_capacity(num_updates * 2);
    let mut t = 0i64;

    for _ in 0..num_updates {
        while let Some(&(added_at, (u, v))) = expirations.front() {
            if t - added_at < lifetime_steps {
                break;
            }

            expirations.pop_front();
            let key = ordered_edge(u, v);
            if let Some(weight) = edge_weights.get_mut(&key) {
                let new_weight = (*weight - 1.0).max(0.0);
                *weight = new_weight;
                if new_weight == 0.0 {
                    edge_weights.remove(&key);
                    operations.push(SbmInstruction::Delete { time: t, u, v });
                } else {
                    operations.push(SbmInstruction::SetWeight {
                        time: t,
                        u,
                        v,
                        weight: new_weight,
                    });
                }
            }
        }

        let (u, v) = if rng.random_range(0.0..1.0) < internal_prob {
            pick_internal(&mut rng, &cluster_nodes)
        } else {
            pick_cross(&mut rng, &cluster_nodes)
        };
        if u == v {
            t += 1;
            continue;
        }

        operations.push(SbmInstruction::Insert {
            time: t,
            u,
            v,
            weight_delta: 1.0,
        });
        *edge_weights.entry(ordered_edge(u, v)).or_insert(0.0) += 1.0;
        expirations.push_back((t, (u, v)));

        t += 1;
    }

    let mut cluster_labels = Vec::with_capacity(total_nodes);
    for c in 0..k_clusters {
        cluster_labels.extend(std::iter::repeat_n(c, n_per_cluster));
    }

    Ok(GeneratedSbmCommands {
        nodes,
        operations,
        expected_edges: expected_edges as usize,
        cluster_labels,
    })
}

/// Prepare a deterministic SBM update workload for replay into
/// [`InMemoryUndirectedGraph`].
///
/// `step_size` is a fraction of the generated time span, matching dyn-cc's
/// benchmark convention. For example, `0.1` produces roughly ten replay
/// batches.
pub fn prepare_diff_workload_sbm<T>(
    seed: u64,
    n_per_cluster: usize,
    k_clusters: usize,
    p_internal: f64,
    q_external: f64,
    n_multiplier: usize,
    lifetime_multiplier: f64,
    step_size: f64,
) -> Result<SbmDiffWorkload<T>, SbmWorkloadError>
where
    T: FloatScalar,
    Strict<T>: StrictCarrierOps<Scalar = T> + Copy,
{
    validate_sbm_params(
        n_per_cluster,
        k_clusters,
        p_internal,
        q_external,
        n_multiplier,
        lifetime_multiplier,
        step_size,
    )?;

    let commands = generate_sbm_commands(
        seed,
        n_per_cluster,
        k_clusters,
        p_internal,
        q_external,
        n_multiplier,
        lifetime_multiplier,
    )?;
    let Some(first_time) = commands.operations.first().map(|op| op.time()) else {
        return Err(SbmWorkloadError::InvalidParameter(
            "generated command stream was empty",
        ));
    };
    let Some(last_time) = commands.operations.last().map(|op| op.time()) else {
        return Err(SbmWorkloadError::InvalidParameter(
            "generated command stream was empty",
        ));
    };
    if first_time == last_time {
        return Err(SbmWorkloadError::InvalidParameter(
            "generated command stream has zero duration",
        ));
    }

    let time_span = last_time - first_time;
    let snapshot_step = ((step_size * time_span as f64) as i64).max(1);
    let mut graph = InMemoryUndirectedGraph::<usize, T>::new();
    let mut batches = Vec::new();
    let mut cursor = 0usize;

    let mut snapshot_time = first_time + snapshot_step;
    while snapshot_time < last_time {
        let batch =
            drain_snapshot_batch(&commands.operations, &mut cursor, snapshot_time, &mut graph)?;
        if !batch.edge_ops.is_empty() || !batch.node_ops.is_empty() {
            batches.push(batch);
        }
        snapshot_time += snapshot_step;
    }

    let final_batch =
        drain_snapshot_batch(&commands.operations, &mut cursor, last_time, &mut graph)?;
    if !final_batch.edge_ops.is_empty() || !final_batch.node_ops.is_empty() {
        batches.push(final_batch);
    }

    Ok(SbmDiffWorkload {
        nodes: commands.nodes,
        cluster_labels: commands.cluster_labels,
        expected_edges: commands.expected_edges,
        snapshot_step,
        batches,
    })
}

fn drain_snapshot_batch<T>(
    operations: &[SbmInstruction],
    cursor: &mut usize,
    snapshot_time: i64,
    graph: &mut InMemoryUndirectedGraph<usize, T>,
) -> Result<SbmUpdateBatch<T>, SbmWorkloadError>
where
    T: FloatScalar,
    Strict<T>: StrictCarrierOps<Scalar = T> + Copy,
{
    let mut edge_ops = Vec::new();

    while *cursor < operations.len() && operations[*cursor].time() <= snapshot_time {
        let edge_op = instruction_to_edge_op(operations[*cursor], graph)?;
        edge_op.apply_to_graph(graph)?;
        edge_ops.push(edge_op);
        *cursor += 1;
    }

    let node_ops = graph.flush_node_ops();
    Ok(SbmUpdateBatch {
        time: snapshot_time,
        edge_ops,
        node_ops,
    })
}

fn instruction_to_edge_op<T>(
    instruction: SbmInstruction,
    graph: &InMemoryUndirectedGraph<usize, T>,
) -> Result<SbmEdgeOp<T>, SbmWorkloadError>
where
    T: FloatScalar,
    Strict<T>: StrictCarrierOps<Scalar = T> + Copy,
{
    match instruction {
        SbmInstruction::Insert {
            u, v, weight_delta, ..
        } => {
            let old_weight = graph
                .edge_weight(u, v)
                .map(|weight| weight.into_scalar())
                .unwrap_or(T::ZERO);
            let delta = T::from(weight_delta).ok_or(SbmWorkloadError::InvalidWeight)?;
            strict_edge_set(u, v, old_weight + delta)
        }
        SbmInstruction::Delete { u, v, .. } => Ok(SbmEdgeOp::Delete { u, v }),
        SbmInstruction::SetWeight { u, v, weight, .. } => {
            let weight = T::from(weight).ok_or(SbmWorkloadError::InvalidWeight)?;
            strict_edge_set(u, v, weight)
        }
    }
}

fn strict_edge_set<T>(u: usize, v: usize, weight: T) -> Result<SbmEdgeOp<T>, SbmWorkloadError>
where
    T: FloatScalar,
    Strict<T>: StrictCarrierOps<Scalar = T> + Copy,
{
    Ok(SbmEdgeOp::Set {
        u,
        v,
        weight: Strict::<T>::from_positive_scalar(weight)
            .map_err(|_| SbmWorkloadError::InvalidWeight)?,
    })
}

fn validate_sbm_params(
    n_per_cluster: usize,
    k_clusters: usize,
    p_internal: f64,
    q_external: f64,
    n_multiplier: usize,
    lifetime_multiplier: f64,
    step_size: f64,
) -> Result<(), SbmWorkloadError> {
    if n_per_cluster == 0 {
        return Err(SbmWorkloadError::InvalidParameter("n_per_cluster"));
    }
    if k_clusters == 0 {
        return Err(SbmWorkloadError::InvalidParameter("k_clusters"));
    }
    if n_multiplier == 0 {
        return Err(SbmWorkloadError::InvalidParameter("n_multiplier"));
    }
    if !(p_internal.is_finite() && p_internal >= 0.0) {
        return Err(SbmWorkloadError::InvalidParameter("p_internal"));
    }
    if !(q_external.is_finite() && q_external >= 0.0) {
        return Err(SbmWorkloadError::InvalidParameter("q_external"));
    }
    if !(lifetime_multiplier.is_finite() && lifetime_multiplier > 0.0) {
        return Err(SbmWorkloadError::InvalidParameter("lifetime_multiplier"));
    }
    if !(step_size.is_finite() && step_size > 0.0) {
        return Err(SbmWorkloadError::InvalidParameter("step_size"));
    }

    Ok(())
}

fn pick_internal(rng: &mut rand::rngs::StdRng, cluster_nodes: &[Vec<usize>]) -> (usize, usize) {
    let c = rng.random_range(0..cluster_nodes.len());
    let cluster = &cluster_nodes[c];
    if cluster.len() == 1 {
        return (cluster[0], cluster[0]);
    }
    let a = rng.random_range(0..cluster.len());
    let mut b = rng.random_range(0..cluster.len());
    while b == a {
        b = rng.random_range(0..cluster.len());
    }
    (cluster[a], cluster[b])
}

fn pick_cross(rng: &mut rand::rngs::StdRng, cluster_nodes: &[Vec<usize>]) -> (usize, usize) {
    if cluster_nodes.len() < 2 {
        return pick_internal(rng, cluster_nodes);
    }

    let c1 = rng.random_range(0..cluster_nodes.len());
    let mut c2 = rng.random_range(0..cluster_nodes.len());
    while c2 == c1 {
        c2 = rng.random_range(0..cluster_nodes.len());
    }
    let u_cluster = &cluster_nodes[c1];
    let v_cluster = &cluster_nodes[c2];
    (
        u_cluster[rng.random_range(0..u_cluster.len())],
        v_cluster[rng.random_range(0..v_cluster.len())],
    )
}

fn ordered_edge(u: usize, v: usize) -> (usize, usize) {
    if u <= v { (u, v) } else { (v, u) }
}

File: experiments/dynamic_sbm/workloads.py
from __future__ import annotations

import argparse
import math
from collections import deque
from dataclasses import dataclass

import numpy as np


@dataclass(frozen=True)
class UpdateBatch:
    time: int
    edge_ops: list[tuple[int, int, float | None]]


@dataclass(frozen=True)
class SbmWorkload:
    cluster_labels: list[int]
    expected_edges: int
    snapshot_step: int
    batches: list[UpdateBatch]


def prepare_diff_workload_sbm(
    *,
    seed: int,
    n_per_cluster: int,
    k_clusters: int,
    p_internal: float,
    q_external: float,
    n_multiplier: int,
    lifetime_multiplier: float,
    step_size: float,
) -> SbmWorkload:
    validate_sbm_params(
        n_per_cluster,
        k_clusters,
        p_internal,
        q_external,
        n_multiplier,
        lifetime_multiplier,
        step_size,
    )

    total_nodes = n_per_cluster * k_clusters
    expected_internal = (
        k_clusters * n_per_cluster * (n_per_cluster - 1) * 0.5 * p_internal
    )
    expected_external = (
        k_clusters
        * (k_clusters - 1)
        * 0.5
        * n_per_cluster
        * n_per_cluster
        * q_external
    )
    expected_edges = expected_internal + expected_external
    if not math.isfinite(expected_edges) or expected_edges <= 0.0:
        raise ValueError("expected edge count must be positive")

    rng = np.random.default_rng(seed)
    num_updates = math.ceil(n_multiplier * expected_edges)
    lifetime_steps = max(1, math.ceil(lifetime_multiplier * expected_edges))
    internal_prob = expected_internal / expected_edges
    snapshot_step = max(1, int(step_size * max(1, num_updates - 1)))
    next_snapshot = snapshot_step

    edge_weights: dict[tuple[int, int], float] = {}
    expirations: deque[tuple[int, int, int]] = deque()
    batches: list[UpdateBatch] = []
    batch_ops: list[tuple[int, int, float | None]] = []

    for t in range(num_updates):
        while expirations and t - expirations[0][0] >= lifetime_steps:
            _, u, v = expirations.popleft()
            key = ordered_edge(u, v)
            old_weight = edge_weights.get(key)
            if old_weight is None:
                continue

            new_weight = max(0.0, old_weight - 1.0)
            if new_weight == 0.0:
                del edge_weights[key]
                batch_ops.append((u, v, None))
            else:
                edge_weights[key] = new_weight
                batch_ops.append((u, v, new_weight))

        if rng.random() < internal_prob:
            u, v = pick_internal(rng, n_per_cluster, k_clusters)
        else:
            u, v = pick_cross(rng, n_per_cluster, k_clusters)

        if u != v:
            key = ordered_edge(u, v)
            new_weight = edge_weights.get(key, 0.0) + 1.0
            edge_weights[key] = new_weight
            expirations.append((t, u, v))
            batch_ops.append((u, v, new_weight))

        if t >= next_snapshot:
            if batch_ops:
                batches.append(UpdateBatch(time=t, edge_ops=batch_ops))
                batch_ops = []
            while t >= next_snapshot:
                next_snapshot += snapshot_step

    if batch_ops:
        batches.append(UpdateBatch(time=num_updates - 1, edge_ops=batch_ops))

    return SbmWorkload(
        cluster_labels=[node // n_per_cluster for node in range(total_nodes)],
        expected_edges=int(expected_edges),
        snapshot_step=snapshot_step,
        batches=batches,
    )


def pick_internal(
    rng: np.random.Generator,
    n_per_cluster: int,
    k_clusters: int,
) -> tuple[int, int]:
    cluster = int(rng.integers(0, k_clusters))
    start = cluster * n_per_cluster
    if n_per_cluster == 1:
        return start, start

    a = int(rng.integers(0, n_per_cluster))
    b = int(rng.integers(0, n_per_cluster - 1))
    if b >= a:
        b += 1
    return start + a, start + b


def pick_cross(
    rng: np.random.Generator,
    n_per_cluster: int,
    k_clusters: int,
) -> tuple[int, int]:
    if k_clusters < 2:
        return pick_internal(rng, n_per_cluster, k_clusters)

    c1 = int(rng.integers(0, k_clusters))
    c2 = int(rng.integers(0, k_clusters - 1))
    if c2 >= c1:
        c2 += 1

    u = c1 * n_per_cluster + int(rng.integers(0, n_per_cluster))
    v = c2 * n_per_cluster + int(rng.integers(0, n_per_cluster))
    return u, v


def ordered_edge(u: int, v: int) -> tuple[int, int]:
    return (u, v) if u <= v else (v, u)


def apply_edge_ops_to_weight_map(
    edge_ops: list[tuple[int, int, float | None]],
    edge_weights: dict[tuple[int, int], float],
) -> None:
    for u, v, weight in edge_ops:
        edge = ordered_edge(u, v)
        if weight is None:
            edge_weights.pop(edge, None)
        else:
            edge_weights[edge] = weight


def balanced_query_pairs(
    query_nodes: list[int],
    labels: list[int],
    pair_count: int,
    *,
    rng: np.random.Generator,
) -> tuple[list[tuple[int, int]], list[bool]]:
    by_label: dict[int, list[int]] = {}
    for node in query_nodes:
        by_label.setdefault(labels[node], []).append(node)

    same_labels = [label for label, nodes in by_label.items() if len(nodes) >= 2]
    cross_labels = list(by_label)
    same_target = pair_count // 2
    cross_target = pair_count - same_target

    pairs: list[tuple[int, int]] = []
    pair_labels: list[bool] = []

    for _ in range(same_target):
        if not same_labels:
            break
        label = same_labels[int(rng.integers(0, len(same_labels)))]
        nodes = by_label[label]
        i = int(rng.integers(0, len(nodes)))
        j = int(rng.integers(0, len(nodes) - 1))
        if j >= i:
            j += 1
        pairs.append((nodes[i], nodes[j]))
        pair_labels.append(True)

    for _ in range(cross_target):
        if len(cross_labels) < 2:
            break
        left_label = cross_labels[int(rng.integers(0, len(cross_labels)))]
        right_candidates = [label for label in cross_labels if label != left_label]
        right_label = right_candidates[int(rng.integers(0, len(right_candidates)))]
        left_nodes = by_label[left_label]
        right_nodes = by_label[right_label]
        pairs.append(
            (
                left_nodes[int(rng.integers(0, len(left_nodes)))],
                right_nodes[int(rng.integers(0, len(right_nodes)))],
            )
        )
        pair_labels.append(False)

    return pairs, pair_labels


def query_subset(
    live_nodes: list[int],
    *,
    frac: float,
    min_size: int,
    seed: int,
) -> list[int]:
    query_len = max(min_size, int(len(live_nodes) * frac))
    query_len = min(query_len, len(live_nodes))

    nodes = np.array(live_nodes, dtype=np.int64)
    rng = np.random.default_rng(seed)
    rng.shuffle(nodes)
    return sorted(int(node) for node in nodes[:query_len])


def expected_edges_per_node(
    n_per_cluster: int,
    k_clusters: int,
    p_internal: float,
    q_external: float,
) -> int:
    total_nodes = n_per_cluster * k_clusters
    expected_internal = math.ceil((n_per_cluster - 1) * p_internal)
    expected_external = math.ceil((total_nodes - n_per_cluster) * q_external)
    return max(1, expected_internal + expected_external)


def default_sampling_seeds(num_clusters: int, coreset_size: int) -> int:
    return max(2, min(num_clusters * 8, coreset_size - 1))


def parse_pair_sizes(value: str) -> list[int]:
    sizes = []
    for raw_size in value.split(","):
        raw_size = raw_size.strip()
        if not raw_size:
            continue
        size = int(raw_size)
        if size <= 0:
            raise argparse.ArgumentTypeError("pair sizes must be positive")
        sizes.append(size)
    return sizes


def validate_sbm_params(
    n_per_cluster: int,
    k_clusters: int,
    p_internal: float,
    q_external: float,
    n_multiplier: int,
    lifetime_multiplier: float,
    step_size: float,
) -> None:
    if n_per_cluster <= 0:
        raise ValueError("n_per_cluster must be positive")
    if k_clusters <= 0:
        raise ValueError("num_clusters must be positive")
    if n_multiplier <= 0:
        raise ValueError("n_multiplier must be positive")
    if not math.isfinite(p_internal) or p_internal < 0.0:
        raise ValueError("p_internal must be finite and non-negative")
    if not math.isfinite(q_external) or q_external < 0.0:
        raise ValueError("q_external must be finite and non-negative")
    if not math.isfinite(lifetime_multiplier) or lifetime_multiplier <= 0.0:
        raise ValueError("lifetime_multiplier must be positive and finite")
    if not math.isfinite(step_size) or step_size <= 0.0:
        raise ValueError("step_size must be positive and finite")
